### ISSUE TITLE
Question about Firmware Transparency Design: update client (flash_tool) verifying consistency

### ISSUE THREAD

**Issue by Ruide** [google/trillian-examples#221]
Morning!

TLDR' I think we might want FT Log to respond with an inclusion promise (https://github.com/google/trillian/blob/master/docs/TransparentLogging.md#inclusion-proofs-vs-promises) called signed firmware timestamp (SFT similar to SCT in https://www.certificate-transparency.org/how-ct-works) to submission of valid firmware. And then for Update Client, verification of the SFT should be good enough instead of verification of consistency. Verification of consistency leads to availability issues and potential attack surface with URL of FT Log alone. 

The following are the details. Because we are discussing the design, so I will use the language in the Claimant models.

From the description of the overview figure of the design page, the Update Client in the Device block is acting as the Believer role of both FIRMWARE and FIRMWARE_Log. And it is also acting as the Verifier role of FIRMWARE_Log by asking for consistency proof from FT Log and verifies the received proof. This is a 'Verify before Use' mode of believer as mentioned in https://github.com/google/trillian/blob/master/docs/claimantmodel/CoreModel.md#trust-but-verify. And Verify before Use mode is subject to availability issue. While Transparency is mostly used with the 'Trust but Verify' mode AFAIK. And also, for Update Client to verify the identity of FT Log, a URL might not be good enough (DNS poisoning etc). A fake FT log can lead to a Split view attack. So most likely a hard-coded certificate of FT Personality is required. On the other hand, if we let FT Log to respond with a signed firmware timestamp (SFT) to submission of valid firmware to FT Log and let Update Client verifies the SFT. We can solve the availability issue, the potential attack surface and can still achieve eventual discoverability. (Note that, in the current demo implementation of FT Log, publisher.go would submit firmware image and wait for proof of inclusion. Only after the publisher validates the inclusion of proof, would it proceed. So in this case, the MMD field in SFT could be deprecated. And signed firmware timestamp (SFT) might be better names signed firmware promise (SFP)? Just random thoughts.)

Since Update Client is a Believer of FIRMWARE_Log, so it might want to just believe FT_log (by verifying SFT) instead of still acting as Verifier. And since FIRMWARE vendor's monitor and Observer's FT Monitor are already acting as Verifier, so we should be able to assume any inconsistency of FT Log would be reported to Arbiter once found. 

To put it simple, the Update Client needs to validate the blessing from two entities, one is the Vendor and the other is FT Log. And it shall not be the Update Client's duty to make sure these two entities are still working fine. It's FT Log's job to keep an eye on Vendor. And its Log monitor's (could be held by Vendor or third parties) job to watch FT Log. Putting the validation of FT Log working fine burden on Update Client does not seem to add much value. Since if the Vendor is benign, it would notice the misbehavior of FT Log and switch to another benign FT log source before it publishes an image. 

If we want to defend against compromised FIRMWARE Vendor and compromised FT Log colluding together, then we need to add inclusion proof validation in Update Client and add multiple FT Log sources. See https://tools.ietf.org/html/draft-ietf-trans-threat-analysis-16#section-3.3.3. But this is out of scope for this issue.

Please let me know if I have any misunderstandings or incorrect assumptions!

**Comment by AlCutter** [comment]
Hi Ruide,

Many thanks for the questions!
There are a few different points in there, I'll try to separate them out and reply to each.

SCTs in general are a much weaker form of "proof" (in reality it's just a promise to integrate), and they don't remove the need for the believer of the statement (the certificate in the case of CT) to check that the promise the SCT represents has been honoured.
You may be interested to know that SCTs did not exist in CT as it was originally designed, and we added them only as a mechanism to ensure that CA issuance latency would be minimally affected - FT does not have this latency constraint; it is unlikely to be a problem if it takes a minute, or even hours, for the log to integrate metadata about new firmware.

The FT update bundle also already has a much stronger statement than an SCT represents: it contains both an STH and inclusion proof and so, more than a promise to integrate, it is a commitment to a log state which already verifiably includes the firmware in question.

As you rightly point out, it _could_ be that this new STH represents a split view, and there are a few things we can do about that, but the first step is have every entity interacting with the log maintain a copy of the latest STH it trusts from a log and only update that to a newer STH if it can be shown to be consistent with the entity's view of the world - in effect this means that any split views have to be maintained by the log indefinitely, this is doable but is tricky and increases the likelihood of detection.

Currently, this is what the `flash_tool` does. 

There are other things we can do to further tighten this to _"is the new log consistent with other people's view of the world?"_ though:
- require that firmware images are logged to N separate logs (this is what CT UA policy currently does), and works on the assumption that at least one of those N logs is honest.
- provide a mechanism whereby the different entities in the system share their views of the log

The second of these is something we're hoping to add to the FT demo next year, we call it `STH Witness` and you can read a bit about what that looks like [here](https://arxiv.org/pdf/2011.04551.pdf) along with security proofs.

Your observation about availability is good - a goal for the demo was to keep it as small and simple as possible to help in understanding, but in a "real world" system there are a number of things we can do:
 - multiple redundant logs helps immensely (the low bar of 99% availability over 90 days in CT policy quickly improves when you have multiple independent logs in the system; for N independent logs systematic availability is 1-(1-0.99)<sup>N</sup>), and having redundancy is a good idea anyway - accidents happen :)
 - we can serve tree internal nodes as static "tiles" representing groups of nodes (this is in fact how Trillian stores the tree data internally, and is how golang module transparency works too), clients can fetch just the tiles they need in order to build their own proofs. Since these tiles are static, they can be cached and served by very simple/scalable HTTP servers
 - depending on the use case, the consistency proof could be retrieved earlier in the process and reused across multiple target devices - e.g. in an IoT/enterprise scenario, you could imagine that the expected "current STH" of the entire fleet is already known ahead of time, so the update bundle could contain the appropriate consistency proof allowing the device to verify both inclusion and consistency off-line. (Similarly, this concept can be extended to cover the addition of STH witness too.)

Hope that helps!

**Comment by Ruide** [comment]
Hi Al,

Much appreciated for your prompt and insightful response! I now understand where I got wrong and the reasoning behind the overall design.

After reading your responses, I found my main misunderstanding sits in my incorrect assumption on Update Client. I thought Update Client should be a process run on an energy/computation limited device during booting. But from your responses, I figured that Update Client is actually running on a server (or a smart hub). And this server is responsible for the update of potentially hundreds of devices. And for the case where Update Client is running on a server, it makes sense to perform consistency proof and act as an auditor. Previously, I was thinking of a standalone IoT device performing an OTA update. So, I was thinking that it would be better to leave the auditor role to other entities and only acts as a Believer of FIRMWARE and FIRMWARE Log (aka validate the identity of vendor and FT Log). 

And after reading your replies, I double-checked the design doc and I just noticed that you have made it clear in the design doc that the target device is only Believer and Update Client is both Believer and Verifier... Sorry for missing your point and not getting what you meant right away...

The followings are some notes I made during reading your responses, please feel free to skip.

> SCTs in general are a much weaker form of "proof" (in reality it's just a promise to integrate), and they don't remove the need for the believer of the statement (the certificate in the case of CT) to check that the promise the SCT represents has been honored.
> You may be interested to know that SCTs did not exist in CT as it was originally designed, and we added them only as a mechanism to ensure that CA issuance latency would be minimally affected - FT does not have this latency constraint; it is unlikely to be a problem if it takes a minute, or even hours, for the log to integrate metadata about new firmware.
> 

Good to know that CT did not have SCT in the beginning. And yes, a signed promise provides a weaker form of trust. Thanks for sharing! 

> The FT update bundle also already has a much stronger statement than an SCT represents: it contains both an STH and inclusion proof and so, more than a promise to integrate, it is a commitment to a log state which already verifiably includes the firmware in question.
> 

Yes, this was also my point. I thought the proof material embedded in the update bundle should be enough for Update Client (but I was wrong about where Update Client runs). Now, since I know Update Client is actually running on a server, then it makes sense for it to verify the state of a FT log by acting as an auditor to retreive and check the consistency proof. 

> As you rightly point out, it _could_ be that this new STH represents a split view, and there are a few things we can do about that, but the first step is have every entity interacting with the log maintain a copy of the latest STH it trusts from a log and only update that to a newer STH if it can be shown to be consistent with the entity's view of the world - in effect this means that any split views have to be maintained by the log indefinitely, this is doable but is tricky and increases the likelihood of detection.
> 
> Currently, this is what the `flash_tool` does.
> 

Yes, the flash_tool is acting as another auditor. And this audits on the trust state a FT log. 

> There are other things we can do to further tighten this to _"is the new log consistent with other people's view of the world?"_ though:
> 
> * require that firmware images are logged to N separate logs (this is what CT UA policy currently does), and works on the assumption that at least one of those N logs is honest.
> * provide a mechanism whereby the different entities in the system share their views of the log
> 

Yes, this is consistent with the general transparency idea.

> The second of these is something we're hoping to add to the FT demo next year, we call it `STH Witness` and you can read a bit about what that looks like [here](https://arxiv.org/pdf/2011.04551.pdf) along with security proofs.
> 

Combining crypto and concensus in distributed system to collaboratively audit on a log is fun. This greatly fortifies the trust on Log a Log Believer can have. 

> Your observation about availability is good - a goal for the demo was to keep it as small and simple as possible to help in understanding, but in a "real world" system there are a number of things we can do:
> 
> * multiple redundant logs helps immensely (the low bar of 99% availability over 90 days in CT policy quickly improves when you have multiple independent logs in the system; for N independent logs systematic availability is 1-(1-0.99)N), and having redundancy is a good idea anyway - accidents happen :)
> * we can serve tree internal nodes as static "tiles" representing groups of nodes (this is in fact how Trillian stores the tree data internally, and is how golang module transparency works too), clients can fetch just the tiles they need in order to build their own proofs. Since these tiles are static, they can be cached and served by very simple/scalable HTTP servers
> * depending on the use case, the consistency proof could be retrieved earlier in the process and reused across multiple target devices - e.g. in an IoT/enterprise scenario, you could imagine that the expected "current STH" of the entire fleet is already known ahead of time, so the update bundle could contain the appropriate consistency proof allowing the device to verify both inclusion and consistency off-line. (Similarly, this concept can be extended to cover the addition of STH witness too.)
> 

Yes, the first two points are consistent with the current transparency idea. The third point is actually what I was thinking before. And I was thinking about the off-line validation for the device's part. Now that I understand Update Client is not running on the device, the whole design makes sense to me!

Thanks again for your detailed clarifications! They are crystal clear and help keeping it straight in my mind!
